Add qd8-f16/f32-qb4w mlal kernels

GregoryComer opened the pull request:
Kernels, tests, and benchmarks for qd8-f16/f32-qb4w neon mlal kernels.

TODO Perf numbers, remove k epilogue?

GregoryComer committed:
Functional qd8-f32-qb4w avx2 1x8c8 kernel

GregoryComer committed:
Add metakernel and tests for qd8-f16-qb4w avx2

GregoryComer committed:
Benchmark support for qb4w avx2

GregoryComer committed:
Add QB4W MLAL metakernels

digantdesai committed:
[qb4w] Add Neoni8mm kernels

* F16 and F32 gemm kernels
* Tests and benchmarks

digantdesai commented on src/qd8-f32-qb4w-gemm/gen/qd8-f32-qb4w-gemm-1x16-minmax-neon-mlal-lane-prfm.c:
```diff
@@ -0,0 +1,334 @@
+#include <xnnpack/prefetch.h>
+
+
+void xnn_qd8_f32_qb4w_gemm_minmax_ukernel_1x16__neon_mlal_lane_prfm(
```
NR=16 - I would be curious to see how does this perform on smaller cores. Can you please benchmark it? I am particularly interested in seeing how does this i.e. QB4W MRx16x8 compare against QC4W MRx8x16.

digantdesai commented on src/qd8-f32-qb4w-gemm/gen/qd8-f32-qb4w-gemm-6x16-minmax-neon-mlal-lane.c:
```diff
@@ -0,0 +1,930 @@
+      const int8x8_t vb89ABCDEFc6 = vshl_n_s8(vb89ABCDEFc67, 4);
+      const int8x8_t vb89ABCDEFc7 = vand_s8(vb89ABCDEFc67, vmask);
+
+      const int16x8_t vxb01234567c0 = vmovl_s8(vb01234567c0);
```
interleave these widening with the load to reduce register pressure

digantdesai commented on src/qd8-f32-qb4w-gemm/gen/qd8-f32-qb4w-gemm-6x16-minmax-neon-mlal-lane.c:
```diff
@@ -0,0 +1,930 @@
+      const int16x8_t vxb89ABCDEFc7 = vmovl_s8(vb89ABCDEFc7);
+
+
+      vacc0x0123 = vmlal_lane_s16(vacc0x0123, vget_low_s16(vxb01234567c0), vget_low_s16(vxa0), 0);
```
tell me how much spill with this large tile size IIRC this is what they have in the gemm-config?

digantdesai commented on src/qs8-gemm/neon-mlal-lane.c.in:
```diff
@@ -318,39 +347,81 @@ void xnn_${DATATYPE_SPEC}_gemm_minmax${REQUANTIZATION_SPEC}_ukernel_${MR}x${NR}_
-      $for M in range(0, MR, 2):
-        $if M + 1 == MR:
-          const float32x4_t vinput_scale${M} = vld1q_dup_f32(&quantization_params[${M}].inv_scale);
+        $for N in range(0, NR, 4):
```
what if you flop the loop in the generator (metakernel), does it help with the register pressure?
i.e.
`
for(MR)
  for(NR)
`
vs.
`
for(NR)
  for(MR)
`  
also you can interleave the scales (which has long life right now) to see if it helps.

GregoryComer committed:
Fix qb4w avx2 microkernel indentation

digantdesai committed:
[QB4W] neoni8mm: fix indent

xnnpack-bot committed:
Merge pull request #6681 from digantdesai:qb4w_neoni8mm

PiperOrigin-RevId: 650515334

gonnet committed:
Add subgraph support for `qp8_f32_qc4w` GEMMs.

PiperOrigin-RevId: 650552984

alankelly committed:
Merge branch 'master' into qb4w-avx2

alankelly committed:
add scalar exp microkernels

PiperOrigin-RevId: 650582656

alankelly committed:
f32 exp operator

PiperOrigin-RevId: 650625384

alankelly committed:
Add Exp Subgraph API

PiperOrigin-RevId: 650641360

qukhan committed:
Don't use `bazel_plaforms` to detect wasm builds.

We get into weird errors that look a lot like [this][1].

[1]: https://github.com/bazelbuild/bazel/issues/18567#issuecomment-1575276177

PiperOrigin-RevId: 650665259

fbarchard committed:
AVX10 VNNI qs8-rsum microkernels

- add avx2 and avx10 vnni microkernels
- build support for avx256vnni and avx256vnnigfni

PiperOrigin-RevId: 650709970

gonnet committed:
Add an internal flag to coerce `f32_qd8` conversions to `f32_qp8` in `xnn_define_convert()`.

PiperOrigin-RevId: 650713716

fbarchard committed:
Run templates to generate build files and source

PiperOrigin-RevId: 650762203

xnnpack-bot committed:
Merge pull request #6618 from GregoryComer:qb4w-avx2

PiperOrigin-RevId: 650890388

gonnet committed:
Run templates to generate build files and source

PiperOrigin-RevId: 650899970

xnnpack-bot committed:
Merge pull request #6618 from GregoryComer:qb4w-avx2

PiperOrigin-RevId: 650901182
